tensorflow/compiler/mlir/tools/kernel_gen/transforms/tf_64bit_indexer.cc

```diff
@@ -0,0 +1,206 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include <string>
+#include <utility>
+#include <iostream>
+
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SmallSet.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Dialect/SCF/SCF.h"  // from @llvm-project
+#include "mlir/Dialect/Shape/IR/Shape.h"  // from @llvm-project
+#include "mlir/Dialect/StandardOps/IR/Ops.h"  // from @llvm-project
+#include "mlir/IR/Block.h"  // from @llvm-project
+#include "mlir/IR/BlockAndValueMapping.h"  // from @llvm-project
+#include "mlir/IR/BuiltinAttributes.h"  // from @llvm-project
+#include "mlir/IR/BuiltinOps.h"  // from @llvm-project
+#include "mlir/IR/Location.h"  // from @llvm-project
+#include "mlir/IR/MLIRContext.h"  // from @llvm-project
+#include "mlir/IR/OperationSupport.h"  // from @llvm-project
+#include "mlir/IR/PatternMatch.h"  // from @llvm-project
+#include "mlir/IR/TypeRange.h"  // from @llvm-project
+#include "mlir/Support/LogicalResult.h"  // from @llvm-project
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"  // from @llvm-project
+#include "tensorflow/compiler/mlir/tensorflow/ir/tf_dialect.h"
+#include "tensorflow/compiler/mlir/tools/kernel_gen/ir/tf_framework_ops.h"
+#include "tensorflow/compiler/mlir/tools/kernel_gen/transforms/passes.h"
+#include "tensorflow/compiler/mlir/tools/kernel_gen/transforms/rewriters.h"
+#include "mlir/Transforms/Bufferize.h"  // from @llvm-project
+
+
+namespace mlir {
+namespace kernel_gen {
+namespace transforms {
+namespace {
+
+
+using shape::ShapeOfOp;
+static constexpr StringRef kEmitCInterfaceAttrName = "llvm.emit_c_interface";
+
+bool IsTFOperation(Operation *op) {
+  return op != nullptr &&
+         op->getDialect() ==
+             op->getContext()->getLoadedDialect<TF::TensorFlowDialect>();
+}
+
+struct TFT64BitIndexerPattern : public RewritePattern {
+  explicit TFT64BitIndexerPattern(MLIRContext *ctx, bool index_64bit)
+      : RewritePattern(MatchAnyOpTypeTag(), /*benefit=*/1, ctx) {index_64bit_ = index_64bit;}
+
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const override {
+    if (!IsTFOperation(op) || op->getParentOfType<tf_framework::JITCompileOp>() ||
+          !llvm::isa<FuncOp>(op->getParentOp()))
+    {
+        return failure();
+    }
+    SmallVector<Operation *, 16> cluster;
```

Comment:
As we wanted to assume that there is only one TF op, we can avoid the clustering entirely. With this assumption there is no need for `cluster`, `operand_set`, and `result_set`. Below, we can then get the operands, results, and types, directly from the op we're looking at `op`.